st2common/st2common/models/db/__init__.py

```diff
@@ -498,6 +498,48 @@ def _process_datetime_range_filters(self, filters, order_by=None):
         return filters, order_by_list
 
 
+class ChangeRevisionMongoDBAccess(MongoDBAccess):
+
+    def get(self, *args, **kwargs):
+        instance = super(ChangeRevisionMongoDBAccess, self).get(*args, **kwargs)
+        instance._original_rev = instance.rev
+        return instance
+
+    def insert(self, instance):
+        instance = self.model.objects.insert(instance)
+        instance._original_rev = instance.rev
+
+        return self._undo_dict_field_escape(instance)
+
+    def add_or_update(self, instance):
+        if not hasattr(instance, '_original_rev'):
+            return self.insert(instance)
+        else:
+            save_condition = {
+                'id': instance.id,
+                'rev': instance._original_rev
+            }
+
+            instance.rev = instance._original_rev + 1
+            instance.save(save_condition=save_condition)
```

Comment:
I wonder if we should catch the error here and reraise it as st2 conflict exception

Reply:
Yeah, I'll rewrap it as a st2 exception.